crates/bevy_ecs/src/reflect/mod.rs

```diff
@@ -59,20 +61,34 @@ fn from_reflect_or_world<T: FromReflect>(
         return value;
     }
 
-    // Clone the `ReflectFromWorld` because it's cheap and "frees"
-    // the borrow of `world` so that it can be passed to `from_world`.
-    let Some(reflect_from_world) = registry.get_type_data::<ReflectFromWorld>(TypeId::of::<T>())
-    else {
+    fn different_type_error<T>(reflected: &str) -> ! {
         panic!(
-            "`FromReflect` failed and no `ReflectFromWorld` registration found for `{}`",
+            "the registration for the reflected `{}` trait for the type `{}` produced \
+            a value of a different type",
+            reflected,
             // FIXME: once we have unique reflect, use `TypePath`.
             std::any::type_name::<T>(),
```

Comment:
Are we not able to fix this now?

Reply:
It could use `TypePath`, but it would have to be added as an additional bound (technically another breaking change) because it is not a supertrait of `Reflect` or `FromReflect`. I wonder though how breaking it would be in practice: isn't `TypePath` derived alongside `Reflect`?

`Reflect` does have a `DynamicTypePath` supertrait, but unfortunately here we don't have an instance of `T` (in fact we just failed to get an instance of `T`) so it cannot be used.

Reply:
True. I suppose it's not really necessary anyways since this is just for a panic message.